pkg/operator/admission/controller.go

```diff
@@ -0,0 +1,118 @@
+type admitFunc func(*v1beta1.AdmissionRequest, *clusterd.Context) error
+
+// doServeAdmitFunc parses the HTTP request for an admission controller webhook, and -- in case of a well-formed
+// request -- delegates the admission control logic to the given admitFunc. The response body is then returned as raw
+// bytes.
+func doServeAdmitFunc(w http.ResponseWriter, r *http.Request, admit admitFunc) ([]byte, error) {
+	// Step 1: Request validation. Only handle POST requests with a body and json content type.
+	if r.Method != http.MethodPost {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return nil, errors.Errorf("invalid method %q, only POST requess are allowed", r.Method)
+	}
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, errors.Wrap(err, "could not read request body")
+	}
+	if contentType := r.Header.Get("Content-Type"); contentType != jsonContentType {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, errors.Errorf("unsupported content type %q, only %q is supported", contentType, jsonContentType)
+	}
+	// Step 2: Parse the AdmissionReview request.
+	var admissionReviewReq v1beta1.AdmissionReview
+	if _, _, err := universalDeserializer.Decode(body, nil, &admissionReviewReq); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, errors.Wrap(err, "failed to deserialize request")
+	} else if admissionReviewReq.Request == nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, errors.New("malformed admission review. request is nil")
+	}
+	// Step 3: Construct the AdmissionReview response.
+	admissionReviewResponse := v1beta1.AdmissionReview{
+		Response: &v1beta1.AdmissionResponse{
+			UID: admissionReviewReq.Request.UID,
+		},
+	}
+	err = admit(admissionReviewReq.Request, rookContext)
+	if err != nil {
+		// If the handler returned an error, incorporate the error message into the response and deny the object
+		// creation.
+		admissionReviewResponse.Response.Allowed = false
+		admissionReviewResponse.Response.Result = &metav1.Status{
+			Message: err.Error(),
+		}
+	} else {
+		admissionReviewResponse.Response.Allowed = true
+	}
+	// Return the AdmissionReview with a response as JSON.
+	bytes, err := json.Marshal(&admissionReviewResponse)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal response to AdmissionReview")
+	}
+	return bytes, nil
+}
+
+// serveAdmitFunc is a wrapper around doServeAdmitFunc that adds error handling and logging.
+func serveAdmitFunc(w http.ResponseWriter, r *http.Request, admit admitFunc) {
+	logger.Info("handling webhook request")
+
+	var writeErr error
+	if bytes, err := doServeAdmitFunc(w, r, admit); err != nil {
+		logger.Errorf("failed to handle webhook request. %v", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		_, writeErr = w.Write([]byte(err.Error()))
+	} else {
+		logger.Info("webhook request handled successfully")
```

Comment:
Can we also log the name of the CR that was processed? Or else it would be good to add logging in another function where the CR is processed.

Reply:
?